csrc/lower_thread_predicate.cpp

```diff
@@ -494,6 +519,224 @@ class RedundantUseAnalysis : BackwardVisitor {
+            merge->inner());
+        TORCH_INTERNAL_ASSERT(
+            outer_iter_im != intermediate_domains.end(),
+            "Couldn't find ",
+            merge->outer());
+        TORCH_INTERNAL_ASSERT(
+            inner_iter_im != intermediate_domains.end(),
+            "Couldn't find ",
+            merge->inner());
+        int dist = std::distance(inner_iter_im, outer_iter_im);
+        if (std::abs(dist) != 1) {
+          // if the merged domains are not adjacent, return empty so the
+          // predication is not applied.
+          return std::vector<IterDomain*>();
+        }
+        // insert after erase so the beg iter is not invalidated
+        auto beg = dist > 0 ? inner_iter_im : outer_iter_im;
+        auto end = dist > 0 ? outer_iter_im : inner_iter_im;
+        intermediate_domains.erase(beg, end + 1);
+        intermediate_domains.insert(beg, merge->out());
+
+        auto outer_iter =
+            std::find(root_domain.begin(), root_domain.end(), merge->outer());
+        auto inner_iter =
+            std::find(root_domain.begin(), root_domain.end(), merge->inner());
+        // outer_iter and inner_iter should be neighbors in root_domain
+        if (inner_iter != root_domain.end() &&
+            outer_iter != root_domain.end() &&
+            std::distance(outer_iter, inner_iter) != 1) {
+          return std::vector<IterDomain*>();
+        }
+        if (inner_iter != root_domain.end()) {
+          merged_root_domains.emplace_back(*inner_iter);
+          index_root_domain.emplace_back(inner_iter - root_domain.begin());
+        }
+        if (outer_iter != root_domain.end()) {
+          merged_root_domains.emplace_back(*outer_iter);
+          index_root_domain.emplace_back(outer_iter - root_domain.begin());
+        }
+      } else {
+        // current analysis of predication is only valid if all the exprs
+        // between this lead domain and root domains are merge
+        return std::vector<IterDomain*>();
+      }
+    }
+    // The following sort is added because in NVFuserTest.FusionIssue2076_CUDA
+    // the order is [B2, I1, I3] while the correct order should be [I3, B2, I1]
+    int n_elements = merged_root_domains.size();
+    std::vector<int> indices(n_elements);
+    std::iota(indices.begin(), indices.end(), 0);
+    std::sort(indices.begin(), indices.end(), [&](int a, int b) {
+      return index_root_domain[a] > index_root_domain[b];
+    });
+    std::vector<IterDomain*> merged_root_domains_sorted(n_elements);
+    for (int i = 0; i < n_elements; ++i) {
+      merged_root_domains_sorted[i] = merged_root_domains[indices[i]];
+    }
+
+    return merged_root_domains_sorted;
+  };
+
+  // Set stride when use extent of merged root domains to index leaf domain.
+  // e.g. Root: [I1,B2,B3] -> Leaf: [I1B2B3], the merged_root_domains =
+  // {B3,B2,I1}. root_stride = {1, len(B3), len(B3) * len(B2)}. For broadcast
+  // root domain, use the extent of its concretized domain.
```

Comment:
@naoyam does a concretized domain mean it's inline concretized? I'm not sure if this PR is assuming that concretized domains it's processing are shared inlined domains of consumers.

Reply:
No, concretization has nothing to do with inlining.